pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+		klog.V(2).InfoS("Processing Node UPDATE event, node cluster status not changed", "nodeName", node.Name)
+		return
+	}
+	failoverPools := oldMatches.Union(newMatches)
+	c.enqueueEIPs(failoverPools)
+	klog.V(2).InfoS("Processing Node UPDATE event", "nodeName", node.Name, "eipNum", failoverPools.Len())
+}
+
+func (c *Cluster) enqueueEIPs(eips sets.String) {
+	for _, eip := range eips.List() {
+		c.queue.Add(eip)
+	}
+}
+
+func (c *Cluster) enqueueExternalIPPool(obj interface{}) {
+	eip, ok := obj.(*v1alpha2.ExternalIPPool)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(typeAssertionErr, "Received unexpected object", "obj", obj)
+			return
+		}
+		eip, ok = deletedState.Obj.(*v1alpha2.ExternalIPPool)
+		if !ok {
+			klog.ErrorS(typeAssertionErr, "DeletedFinalStateUnknown contains non-eip object", "obj", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(eip.Name)
+}
+
+func (c *Cluster) allClusterMembers() (clusterNodes []string) {
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Listing Nodes failed")
+		return
+	}
+	klog.V(3).InfoS("Sync cluster members", "node num", len(nodes))
+
+	for _, node := range nodes {
+		nodeAddr, err := k8s.GetNodeAddr(node)
+		if err != nil {
+			klog.ErrorS(err, "Obtain local IP address from K8s node failed")
+			continue
+		}
+		member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) matchNodeSelector(node *corev1.Node, eip *v1alpha2.ExternalIPPool) bool {
+	nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+	if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+		return true
+	}
+	return false
+}
+
+func (c *Cluster) filterEIPsFromNodeLabels(node *corev1.Node) sets.String {
+	pools := sets.NewString()
+	eips, err := c.externalIPPoolLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Filter EIPs from nodeLabels failed")
+		return pools
+	}
+	for _, eip := range eips {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(eip.Name)
+		}
+	}
+	return pools
+}
+
+func (c *Cluster) members() sets.String {
+	return sets.NewString(c.allClusterMembers()...)
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalIPPoolInformerHasSynced, c.nodeListerSynced) {
+		return
+	}
+
+	defer func() {
+		if err := c.mList.Shutdown(); err != nil {
+			klog.ErrorS(err, "Shut down cluster failed")
+		}
+		close(c.nodeEventsCh)
+	}()
+
+	n, err := c.mList.Join(c.members().List())
+	if err != nil {
+		klog.ErrorS(err, "Join cluster failed", "num", n)
+	}
+	if err := c.initConsistentHash(); err != nil {
```

Comment:
This is not necessary, all pools will be called once by ADD events. I think you can remove the method too.

Reply:
done